File: disk_sense/jobs.py
"""异步操作任务管理器（async-jobs 能力）。

长耗时变更操作（如 17.8GB 删除）改为异步：/operation?async_mode=true
立即返回 job_id，操作在后台线程执行，客户端轮询 GET /job?job_id=。

设计（design.md D2）：
- 进程内内存字典 + 环形上限（默认 1000），不引入外部队列；
- 状态机 pending → running → succeeded/failed；
- 进度由操作函数按批次回调更新（已处理字节/条目）；
- 服务重启后内存丢失：未知 job_id 查询返回「任务已中断」类 failed
  （而非 404），避免客户端永久轮询。
"""

from __future__ import annotations

import threading
import time
import uuid
from collections import deque
from dataclasses import dataclass, field
from typing import Any, Callable, Optional


JOB_PENDING = "pending"
JOB_RUNNING = "running"
JOB_SUCCEEDED = "succeeded"
JOB_FAILED = "failed"
JOB_INTERRUPTED = "interrupted"  # 服务重启后无法追溯的运行中任务

_MAX_JOBS = 1000


@dataclass
class Job:
    """单个异步任务的运行时状态。"""

    job_id: str
    op_type: str
    sources: list[str]
    status: str = JOB_PENDING
    progress: float = 0.0
    processed_items: int = 0
    processed_bytes: int = 0
    total_items: int = 0
    error: Optional[str] = None
    result: Optional[dict] = None
    op_uuid: Optional[str] = None  # 操作审计批次 id（结果摘要关联）
    created_at: float = field(default_factory=time.time)
    finished_at: Optional[float] = None
    lock: threading.Lock = field(default_factory=threading.Lock, repr=False)

    def to_dict(self) -> dict:
        with self.lock:
            d: dict[str, Any] = {
                "job_id": self.job_id,
                "op_type": self.op_type,
                "status": self.status,
                "progress": round(self.progress, 3),
                "processed_items": self.processed_items,
                "processed_bytes": self.processed_bytes,
                "created_at": self.created_at,
            }
            if self.error:
                d["error"] = self.error
            if self.result is not None:
                d["result"] = self.result
            if self.op_uuid:
                d["op_uuid"] = self.op_uuid
            if self.finished_at:
                d["finished_at"] = self.finished_at
            return d


class JobManager:
    """进程内任务注册表（线程安全，环形上限防内存膨胀）。"""

    def __init__(self, max_jobs: int = _MAX_JOBS):
        self._jobs: dict[str, Job] = {}
        self._order: deque[str] = deque()
        self._max_jobs = max_jobs
        self._lock = threading.Lock()

    def create(self, op_type: str, sources: list[str]) -> Job:
        job = Job(job_id=f"job-{uuid.uuid4().hex[:12]}", op_type=op_type, sources=list(sources))
        with self._lock:
            self._jobs[job.job_id] = job
            self._order.append(job.job_id)
            while len(self._order) > self._max_jobs:
                old = self._order.popleft()
                self._jobs.pop(old, None)
        return job

    def get(self, job_id: str) -> Optional[Job]:
        with self._lock:
            return self._jobs.get(job_id)

    def mark_running(self, job: Job) -> None:
        with job.lock:
            if job.status == JOB_PENDING:
                job.status = JOB_RUNNING

    def update_progress(self, job: Job, processed_items: int, processed_bytes: int,
                        total_items: int = 0) -> None:
        with job.lock:
            if job.status != JOB_RUNNING:
                return
            job.processed_items = processed_items
            job.processed_bytes = processed_bytes
            if total_items:
                job.total_items = total_items
                job.progress = min(0.99, processed_items / max(1, total_items))

    def finish(self, job: Job, result: Optional[dict] = None,
               error: Optional[str] = None) -> None:
        with job.lock:
            job.finished_at = time.time()
            if error is not None:
                job.status = JOB_FAILED
                job.error = error
            else:
                job.status = JOB_SUCCEEDED
                job.result = result
                job.progress = 1.0
                if isinstance(result, dict):
                    job.op_uuid = result.get("op_uuid")

    def interrupt_all_running(self) -> int:
        """服务关闭前调用：把运行中任务标记为中断（重启后不可追溯）。"""
        n = 0
        with self._lock:
            jobs = list(self._jobs.values())
        for job in jobs:
            with job.lock:
                if job.status in (JOB_PENDING, JOB_RUNNING):
                    job.status = JOB_INTERRUPTED
                    job.error = "服务重启，任务已中断（操作审计见 /history）"
                    job.finished_at = time.time()
                    n += 1
        return n

File: src/operator/jobs.ts
/**
 * 异步操作任务（async-jobs 能力，无 daemon 磁盘版）。
 *
 * 长耗时变更操作改为异步：CLI 立即返回 job_id 并 spawn 一个 detached
 * 子进程执行，进度/结果实时写入 `<dataHome>/jobs/<job_id>.json`，
 * 客户端用 query_job 轮询读取。
 *
 * 状态机：pending → running → succeeded/failed。
 * 状态全部落盘：进程重启后任务状态仍可追溯；无法追溯的运行中任务由
 * 查询端按 created_at 判定并提示「任务已中断」语义。
 */

import fs from "node:fs";
import path from "node:path";
import { randomUUID } from "node:crypto";
import { jobsDir } from "../config.js";

export const JOB_PENDING = "pending";
export const JOB_RUNNING = "running";
export const JOB_SUCCEEDED = "succeeded";
export const JOB_FAILED = "failed";

export interface JobState {
  job_id: string;
  op_type: string;
  sources: string[];
  /** move/copy 的目标目录 */
  dest?: string | null;
  status: string;
  progress: number;
  processed_items: number;
  processed_bytes: number;
  total_items: number;
  error: string | null;
  result: Record<string, unknown> | null;
  op_uuid: string | null;
  created_at: number;
  finished_at: number | null;
}

function jobFile(jobId: string): string {
  return path.join(jobsDir(), `${jobId}.json`);
}

/** 磁盘任务注册表。 */
export class JobStore {
  constructor(private dir = jobsDir()) {}

  private fileOf(jobId: string): string {
    return path.join(this.dir, `${jobId}.json`);
  }

  create(opType: string, sources: readonly string[], dest?: string | null): JobState {
    fs.mkdirSync(this.dir, { recursive: true });
    const job: JobState = {
      job_id: `job-${randomUUID().slice(0, 12)}`,
      op_type: opType,
      sources: [...sources],
      ...(dest !== undefined ? { dest } : {}),
      status: JOB_PENDING,
      progress: 0,
      processed_items: 0,
      processed_bytes: 0,
      total_items: 0,
      error: null,
      result: null,
      op_uuid: null,
      created_at: Date.now() / 1000,
      finished_at: null,
    };
    this.write(job);
    return job;
  }

  get(jobId: string): JobState | null {
    try {
      return JSON.parse(fs.readFileSync(this.fileOf(jobId), "utf-8")) as JobState;
    } catch {
      return null;
    }
  }

  /** 读 + 原子改写（同进程内使用；子进程写整文件原子替换）。 */
  update(jobId: string, patch: Partial<JobState>): JobState | null {
    const job = this.get(jobId);
    if (!job) return null;
    const next = { ...job, ...patch };
    // result/op_uuid 由 finish 设置时不覆盖
    if (patch.result === undefined && job.result !== null) next.result = job.result;
    this.write(next);
    return next;
  }

  markRunning(jobId: string): void {
    const job = this.get(jobId);
    if (job && job.status === JOB_PENDING) this.write({ ...job, status: JOB_RUNNING });
  }

  setProgress(
    jobId: string,
    processedItems: number,
    processedBytes: number,
    totalItems?: number
  ): void {
    const job = this.get(jobId);
    if (!job || job.status !== JOB_RUNNING) return;
    const next: JobState = {
      ...job,
      processed_items: processedItems,
      processed_bytes: processedBytes,
    };
    if (totalItems && totalItems > 0) {
      next.total_items = totalItems;
      next.progress = Math.min(0.99, processedItems / Math.max(1, totalItems));
    }
    this.write(next);
  }

  finish(jobId: string, result?: Record<string, unknown>, error?: string): void {
    const job = this.get(jobId);
    if (!job) return;
    const next: JobState = {
      ...job,
      finished_at: Date.now() / 1000,
    };
    if (error !== undefined) {
      next.status = JOB_FAILED;
      next.error = error;
    } else {
      next.status = JOB_SUCCEEDED;
      next.result = result ?? null;
      next.progress = 1.0;
      if (result && typeof result === "object" && "op_uuid" in result) {
        next.op_uuid = String(result["op_uuid"]);
      }
    }
    this.write(next);
  }

  private write(job: JobState): void {
    const tmp = `${this.fileOf(job.job_id)}.${process.pid}.tmp`;
    fs.writeFileSync(tmp, JSON.stringify(job, null, 1), "utf-8");
    fs.renameSync(tmp, this.fileOf(job.job_id));
  }
}

/** 把任务状态序列化为对 Agent 的响应（对应 Python Job.to_dict）。 */
export function jobToDict(job: JobState): Record<string, unknown> {
  const d: Record<string, unknown> = {
    job_id: job.job_id,
    op_type: job.op_type,
    status: job.status,
    progress: Math.round(job.progress * 1000) / 1000,
    processed_items: job.processed_items,
    processed_bytes: job.processed_bytes,
    created_at: job.created_at,
  };
  if (job.error) d["error"] = job.error;
  if (job.result !== null) d["result"] = job.result;
  if (job.op_uuid) d["op_uuid"] = job.op_uuid;
  if (job.finished_at) d["finished_at"] = job.finished_at;
  return d;
}
